subsys/net/l2_wifi_if_conn/l2_wifi_conn.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(l2_wifi_mgr_conn);
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/wifi_mgmt.h>
+#include <zephyr/net/net_event.h>
+#include <net/wifi_mgmt_ext.h>
+#include <net/l2_wifi_connect.h>
+#include <zephyr/net/conn_mgr_connectivity.h>
+
+#define WIFI_SHELL_MGMT_EVENTS (NET_EVENT_WIFI_CONNECT_RESULT | \
+				NET_EVENT_WIFI_DISCONNECT_RESULT)
+
+static struct net_mgmt_event_callback net_l2_mgmt_cb;
+int connection_status;
+
+K_SEM_DEFINE(wait_for_conn_event_sem, 0, 1);
+
+int net_l2_wifi_connect(struct conn_mgr_conn_binding *const binding);
+int net_l2_wifi_disconnect(struct conn_mgr_conn_binding *const binding);
+
+/* Delayable work used to handle Wi-Fi connection timeouts. */
+static void wifi_conn_work_handler(struct k_work *work);
+static K_WORK_DELAYABLE_DEFINE(wifi_conn_work, wifi_conn_work_handler);
+
+uint64_t time_stamp_start;
+uint32_t ticks_elapsed;
+
+/* Store a local reference to the connection binding */
+struct conn_mgr_conn_binding  *wifi_conn_binding;
+
+static void wifi_conn_work_handler(struct k_work *work)
+{
+	int rc;
+
+	/* Break if connection was successful */
+	k_sem_take(&wait_for_conn_event_sem, K_NO_WAIT);
+	if (connection_status == WIFI_STATE_COMPLETED) {
+		k_sem_give(&wait_for_conn_event_sem);
+		goto exit;
+	} else {
+		if (!conn_mgr_if_get_flag(wifi_conn_binding->iface, CONN_MGR_IF_PERSISTENT)) {
+			if (connection_status == WIFI_STATE_DISCONNECTED) {
+				net_l2_wifi_disconnect(wifi_conn_binding);
+			}
+			k_sem_give(&wait_for_conn_event_sem);
+			goto exit;
+		} else {
+			k_sem_give(&wait_for_conn_event_sem);
+		}
+	}
+
+	ticks_elapsed = k_ticks_to_ms_ceil32((uint64_t)k_uptime_ticks() - time_stamp_start);
+	if (wifi_conn_binding->timeout && (ticks_elapsed >= ((wifi_conn_binding->timeout)*1000))) {
+		net_l2_wifi_disconnect(wifi_conn_binding);
+		goto exit;
+	}
+
+	rc = net_mgmt(NET_REQUEST_WIFI_CONNECT_STORED, wifi_conn_binding->iface, NULL, 0);
+
+	if (rc) {
+		/*Error in sending net_mgmt command*/
+		return;
+	}
+
+	k_work_reschedule(&wifi_conn_work, K_SECONDS(5));
+	return;
+
+exit:
+	k_work_cancel_delayable(&wifi_conn_work);
```

Comment:
This isn't needed as you are skipping reschedule and also might be racy, esp. without checking return values: https://docs.zephyrproject.org/3.1.0/kernel/services/threads/workqueue.html#check-return-values